Genome graphs have emerged as a powerful framework for the discovery and genotyping of genetic variation, especially notable with respect of complex regions and structural variations in large cohorts; however, their widespread adoption remains limited by a lack of well-established and user-friendly tools and pipelines for their application.
We present results from sub-project "B" of the project "GraphGenomes", the aim of which is to unlock the potential of graph-based genome analysis based on short read data, the most widely employed sequencing technology, for a broad user base. Specifically, we present (i) a standardized workflow for reference graphs construction and maintenance, based on an implementation the NovoGraph algorithm in a scalable and user-friendly Nextflow workflow; (ii) a pipeline for the application and benchmarking of state-of-the-art graph-genome-based approaches; (iii) integration of graph-genome-based inference into the German bioinformatics infrastructure de.NBI. Pipeline implementation was carried out using Snakemake; the developed pipelines include steps for downloading all required input data, for the application of various state-of-the-art graph genome inference frameworks (GraphTyper, BayesTyper and the "PanGenie" algorithm, developed by GraphGenomes sub-project "A"), and for the computation of relevant accuracy metrics (weighted genotype concordance, precision, recall and F-score) against ground truth genotype sets. de.NBI deployment and configuration is based on Ansible and enables the automated installation of relevant pangenomics software within the virtual machines of the de.NBI cloud infrastructure. The implemented workflows and pipelines are thoroughly tested inside the deNBI resources and made available as open-source to further enable community uptake.
We demonstrate the application of the developed components by carrying out an automated evaluation of PanGenie against GraphTyper and BayesTyper based on genome graphs provided by the "Human Genome Structural Variation Consortium" (HGSVC) on a standard benchmarking sample of the "Genome In A Bottle" (GIAB) consortium. PanGenie outperforms GraphTyper for all variant sizes and genotypes a much larger fraction of variants not typable by BayesTyper, while achieving similar performance. PanGenie also outperforms both methods with respect to F-score, with particularly pronounced performance benefits for  large variants (>= 50bp).
In summary, our work reduces the technical hurdles and computational constraints that currently limit the use of graph genome solutions by providing scalable and reproducible workflows that can be deployed in the German bioinformatics infrastructure.
